app/models/king.rb

```diff
@@ -1,19 +1,56 @@
 # rubocop:disable Metrics/AbcSize
 # rubocop:disable Style/SymbolProc
 class King < Piece
+  @rook_id = nil # used by piece to get castled rook ID
+
+  # ensure King didn't move itself into check
+  def finalize_move!(to_row, to_col)
+    return false if get_attackers(to_row, to_col).present?
+    move_to!(to_row, to_col)
+  end
+
   def move_legal?(to_row, to_col)
+    @rook_id = nil
     row_diff = (row - to_row).abs
     col_diff = (col - to_col).abs
-    row_diff <= 1 && col_diff <= 1
-    # add castling moves to the 'whitelist', then verify the validity of the castle
-    #   further down in the logic
+    return true if row_diff <= 1 && col_diff <= 1
+
+    # King castling?
+    return false unless row_diff.zero? && col_diff == 2
+    move_legal_castle?(to_col)
   end
 
-  # override to enable castling
-  def finalize_move!(to_row, to_col)
-    return false if get_attackers(to_row, to_col).present?
-    move_to!(to_row, to_col)
-    # for castling something like: move_the_rook if castling?
+  def move_obstructed?(_to_row, _to_col)
+    false
+  end
+
+  def castling?(to_row, to_col)
+    return "castle. rook: #{@rook_id}" if row == to_row && (col - to_col).abs == 2
```

Comment:
The use of an instance variable is okay. But it's not necessary to record the rook's id; we wouldn't be using it in any logic. Usually the castling direction is identified by 'kingside' (king moves away from initial queen position) or 'queenside' (king moves toward initial queen position).